### PR TITLE
Adding a code action for simplifying the FQN

### LINKED ISSUE #1258: Suggestions to Simplify FQN
Import the class when use full FQN

from
```php
<?php

namespace Foobar\Baz;

class Foo
{
    public function __construct(\Foobar\B<>ar $bar)
    {
    }
}
```

to

```php
<?php

namespace Foobar\Baz;

use Foobar\Bar;

class Foo
{
    public function __construct(Bar $bar)
    {
    }
}
```

### PR DESCRIPTION
**Pull request by mamazu**
Closes #1258

Implementing a code action that replaces the FQN with the class name and an import

### CODE REVIEW

File: lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseSimplifyClassName.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        $sourceBuilder = $this->factory->fromSource($sourceCode);
+        $sourceBuilder->use((string) $type->name());
+        $sourceBuilder->build();
+
+        $newClassName = $type->name()->short();
+        $position = $symbolContext->symbol()->position();
+
+        return new TextDocumentEdits($sourceCode->uri(), TextEdits::fromTextEdits([
+            TextEdit::create($position->start(), $position->end() - $position->start(), $newClassName)
+        ]));
```
Comment (mamazu) [inline lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseSimplifyClassName.php:54]: I don't know how to merge the content of the `$sourceBuilder` and the text edit I made for updating the token with the FQN. Could you maybe give me a hint for that.

Comment (dantleech) [reply lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseSimplifyClassName.php:54]: you need to use the `Updater`, f.e. https://github.com/phpactor/phpactor/blob/07e39698b42f1a146341a8dd62eec7b76e8d7876/lib/CodeTransform/Adapter/TolerantParser/Refactor/TolerantImportName.php#L152

**Comment by dantleech** [comment]
also maye worth renaming everything: https://blog.jetbrains.com/phpstorm/2019/04/reworked-imports-code-cleanup/ this is known as "replace qualifier with import" in PHPStorm I think.

File: lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseSimplifyClassName.php (hunk 1 above)
Comment (mamazu) [reply lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseSimplifyClassName.php]: Yeah of course. Thanks.

**Comment by mamazu** [comment]
I have renamed the code action and the classes accordingly.

**Comment by mamazu** [comment]
@dantleech Anything else needs changing?

**Comment by dantleech** [comment]
sorry, just need to test it locally, will get round to it soon

**Comment by mamazu** [comment]
Ah sure, no rush.

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/ReplaceQualifierWithImportTest.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            SourceCode::fromStringAndPath($source, 'file:///source'),
+            $offset
+        );
+        $sourceCode = SourceCode::fromStringAndPath($source, 'file:///source');
+        $transformed = SourceCode::fromStringAndPath(
+            (string) $textDocumentEdits->textEdits()->apply($sourceCode),
+            $textDocumentEdits->uri()->path()
+        );
+
+        $this->assertEquals(trim($expected), trim($transformed));
```
Comment (dantleech) [inline lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/ReplaceQualifierWithImportTest.php:35]: doesn't matter but use `self::` for `assert`

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/ReplaceQualifierWithImportTest.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        [$source, $expected, $offset] = $this->sourceExpectedAndOffset(__DIR__ . '/fixtures/' . $test);
+
+        $ReplaceQualifierWithImport = new WorseReplaceQualifierWithImport(
+            $this->reflectorForWorkspace($source),
+            new WorseBuilderFactory($this->reflectorForWorkspace($source)),
+            $this->updater()
+        );
+
+        $textDocumentEdits = $ReplaceQualifierWithImport->getTextEdits(
```
Comment (dantleech) [inline lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/ReplaceQualifierWithImportTest.php:25]: there is a `$` here?

**Comment by dantleech** [comment]
the provider is not wired up in the DI container. it's a good idea to test these features locally as they rarely work as expected.

**Comment by mamazu** [comment]
Yeah, testing it out locally is indeed a good idea. I also found another error when testing. But it should work now.

**Comment by dantleech** [comment]
great, thanks :tada:  updated the changelog.

**Pull request merged**